judge/views/contests.py

```diff
@@ -446,39 +450,59 @@ def best_solution_state(points, total):
 def base_contest_ranking_list(contest, problems, queryset, for_user=None):
     cursor = connection.cursor()
     cursor.execute('''
-        SELECT part.id, cp.id, prob.code, MAX(cs.points) AS best, MAX(sub.date) AS `last`
+        SELECT part.id, cp.id, prob.code, MAX(cs.points) AS best, MAX(sub.date) AS `last`,
+             CAST(SUM(CASE WHEN sub.result <> 'AC' AND (sub.result <> 'CE' OR c.time_penalty_is_strict)
+                        THEN c.time_penalty_incorrect_solution
+                        ELSE 0
+                      END) AS INT) AS time_penalty
         FROM judge_contestproblem cp CROSS JOIN judge_contestparticipation part INNER JOIN
              judge_problem prob ON (cp.problem_id = prob.id) LEFT OUTER JOIN
              judge_contestsubmission cs ON (cs.problem_id = cp.id AND cs.participation_id = part.id) LEFT OUTER JOIN
-             judge_submission sub ON (sub.id = cs.submission_id)
-        WHERE cp.contest_id = %s AND part.contest_id = %s {extra}
+             judge_submission sub ON (sub.id = cs.submission_id) LEFT OUTER JOIN
+             judge_contest c ON (cp.contest_id = c.id)
+        WHERE cp.contest_id = %s AND part.contest_id = %s AND sub.status <> 'IE' {extra}
         GROUP BY cp.id, part.id
     '''.format(extra=('AND part.user_id = %s' if for_user is not None else
                                          'AND part.virtual = 0')),
                    (contest.id, contest.id) + ((for_user,) if for_user is not None else ()))
-    data = {(part, prob): (code, best, last and from_database_time(last)) for part, prob, code, best, last in cursor}
+    data = {(part, prob): (code, best, last and from_database_time(last), num_attempts) for part, prob, code, best, last, num_attempts in cursor}
     cursor.close()
 
     problems = map(attrgetter('id', 'points', 'is_pretested'), problems)
 
     def make_ranking_profile(participation):
         part = participation.id
-        return make_contest_ranking_profile(participation, [
-            BestSolutionData(code=data[part, prob][0], points=data[part, prob][1],
-                             time=data[part, prob][2] - participation.start,
-                             state=best_solution_state(data[part, prob][1], points),
-                             is_pretested=is_pretested)
-            if (part, prob) in data and data[part, prob][1] is not None else None
-            for prob, points, is_pretested in problems])
+        contest_progress = []
+        for prob, points, is_pretested in problems:
+            if (part, prob) in data and data[part, prob][1] is not None:
+                best_points = data[part, prob][1]
+                last_attempt_time = data[part, prob][2] - participation.start
+                time_penalty = timedelta(microseconds=data[part, prob][3])
+
+                sol_data =  BestSolutionData(code=data[part, prob][0],
```

Comment:
Two spaces.

Reply:
Can we also put all the contents on a new line so they are only indented with 4 spaces?